File: agents/orchestrator/orchestrator.py
from __future__ import annotations

import asyncio
import json
import os
import socket
from dataclasses import dataclass
from datetime import datetime, timezone

from shared.redis_client import RedisCoordinator
from shared.protocol import AgentMessage, WorkerRequestResponse

import docker


@dataclass(frozen=True)
class OrchestratorSettings:
    redis_url: str
    agents_image: str
    docker_network: str
    repo_volume: str
    max_workers_per_job: int
    cleanup_delay_seconds: float

    @staticmethod
    def from_env() -> "OrchestratorSettings":
        return OrchestratorSettings(
            redis_url=os.environ.get("REDIS_URL", "redis://redis:6379/0"),
            agents_image=os.environ.get("AGENTS_IMAGE", "agentfleet-agents:latest"),
            docker_network=os.environ.get("DOCKER_NETWORK", "agentfleet_default"),
            repo_volume=os.environ.get("REPO_VOLUME", "agentfleet_repo"),
            max_workers_per_job=int(os.environ.get("MAX_WORKERS_PER_JOB", "6")),
            cleanup_delay_seconds=float(os.environ.get("CLEANUP_DELAY_SECONDS", "0")),
        )


class Orchestrator:
    def __init__(self, settings: OrchestratorSettings):
        self.settings = settings
        self.coordinator = RedisCoordinator(settings.redis_url)
        self.docker_client = docker.DockerClient(base_url="unix:///var/run/docker.sock")
        self.hostname = socket.gethostname()

    async def start(self) -> None:
        await self.coordinator.ensure_consumer_group("orchestrators")

        # Launch the worker-request listener alongside the job stream consumer.
        asyncio.create_task(self._worker_request_loop())

        while True:
            messages = await self.coordinator.read_jobs(
                "orchestrators", consumer=f"orch-{self.hostname[:8]}", block_ms=5000
            )
            if not messages:
                continue

            for _stream, entries in messages:
                for entry_id, data in entries:
                    raw = data.get(b"data")
                    if not raw:
                        await self.coordinator.ack_job("orchestrators", entry_id)
                        continue

                    try:
                        job = json.loads(raw.decode("utf-8"))
                    except Exception:
                        await self.coordinator.ack_job("orchestrators", entry_id)
                        continue

                    try:
                        await self._spawn_leader(job)
                        await self.coordinator.ack_job("orchestrators", entry_id)
                    except Exception as e:
                        job_id = str(job.get("job_id") or job.get("id") or "")
                        if job_id:
                            await self.coordinator.publish_job_event(
                                job_id,
                                {"type": "orchestrator_error", "error": str(e)},
                            )
                        await self.coordinator.ack_job("orchestrators", entry_id)

    async def _spawn_leader(self, job: dict) -> None:
        job_id = str(job.get("job_id") or job.get("id") or "")
        if not job_id:
            raise RuntimeError("job_id missing")

        prompt = str(job.get("prompt") or "")

        await self.coordinator.upsert_job(
            job_id,
            {
                "job_id": job_id,
                "prompt": prompt,
                "status": "queued",
                "orchestrated": True,
            },
        )

        leader_id = f"leader-{job_id[:8]}"

        print(
            f"[orchestrator] spawning leader for job_id={job_id}",
            flush=True,
        )

        base_env = {
            "REDIS_URL": self.settings.redis_url,
            "AGENT_SDK_MODE": os.environ.get("AGENT_SDK_MODE", "live"),
            "ANTHROPIC_API_KEY": os.environ.get("ANTHROPIC_API_KEY", ""),
            "MODEL": os.environ.get("MODEL", ""),
            "DRY_RUN_SLEEP_SECONDS": os.environ.get("DRY_RUN_SLEEP_SECONDS", "0"),
            "JOB_ID": job_id,
            "JOB_MODE": "ephemeral",
            "MAX_WORKERS_PER_JOB": str(self.settings.max_workers_per_job),
        }

        self._run_container(
            name=f"agentfleet-{job_id[:8]}-leader",
            env={
                **base_env,
                "AGENT_ROLE": "leader",
                "AGENT_ID": leader_id,
            },
            labels={"agentteam.job_id": job_id, "agentteam.role": "leader"},
        )

        await self.coordinator.publish_job_event(
            job_id,
            {
                "type": "leader_spawned",
                "job_id": job_id,
                "leader_id": leader_id,
            },
        )

        # Background cleanup: when job completes, stop/remove crew containers.
        asyncio.create_task(self._cleanup_when_done(job_id))

    async def _worker_request_loop(self) -> None:
        """Long-running loop: receives WorkerRequests from leaders, spawns workers."""
        while True:
            try:
                request = await self.coordinator.wait_for_worker_request(timeout=5)
                if not request:
                    continue

                job_id = request.job_id
                leader_id = request.leader_id
                requested = request.requested_count

                # Enforce cap.
                granted = min(requested, self.settings.max_workers_per_job)

                print(
                    f"[orchestrator] spawning {granted} workers for job_id={job_id} "
                    f"(requested={requested})",
                    flush=True,
                )

                base_env = {
                    "REDIS_URL": self.settings.redis_url,
                    "AGENT_SDK_MODE": os.environ.get("AGENT_SDK_MODE", "live"),
                    "ANTHROPIC_API_KEY": os.environ.get("ANTHROPIC_API_KEY", ""),
                    "MODEL": os.environ.get("MODEL", ""),
                    "DRY_RUN_SLEEP_SECONDS": os.environ.get(
                        "DRY_RUN_SLEEP_SECONDS", "0"
                    ),
                    "JOB_ID": job_id,
                    "JOB_MODE": "ephemeral",
                }

                worker_ids: list[str] = []
                for i in range(granted):
                    worker_id = f"worker-{job_id[:8]}-{i}"
                    worker_ids.append(worker_id)
                    self._run_container(
                        name=f"agentfleet-{job_id[:8]}-worker-{i}",
                        env={
                            **base_env,
                            "AGENT_ROLE": "worker",
                            "AGENT_ID": worker_id,
                        },
                        labels={
                            "agentteam.job_id": job_id,
                            "agentteam.role": "worker",
                        },
                    )

                # Send response to leader's inbox.
                response = WorkerRequestResponse(
                    job_id=job_id,
                    granted_count=granted,
                    worker_ids=worker_ids,
                    timestamp=datetime.now(timezone.utc).isoformat(),
                )
                msg = AgentMessage(
                    from_agent="orchestrator",
                    text=response.model_dump_json(),
                    summary=f"Granted {granted} workers",
                )
                await self.coordinator.send_message(job_id, leader_id, msg)

                await self.coordinator.publish_job_event(
                    job_id,
                    {
                        "type": "workers_spawned",
                        "job_id": job_id,
                        "leader_id": leader_id,
                        "granted": granted,
                        "worker_ids": worker_ids,
                    },
                )

            except Exception as e:
                print(f"[orchestrator] worker_request_loop error: {e}", flush=True)
                await asyncio.sleep(1)

    def _run_container(
        self, *, name: str, env: dict[str, str], labels: dict[str, str]
    ) -> None:
        image = self.settings.agents_image
        volumes = {self.settings.repo_volume: {"bind": "/repo", "mode": "rw"}}
        try:
            existing = self.docker_client.containers.get(name)
            try:
                existing.remove(force=True)
            except Exception:
                pass
        except Exception:
            pass

        self.docker_client.containers.run(
            image=image,
            name=name,
            detach=True,
            environment=env,
            network=self.settings.docker_network,
            volumes=volumes,
            labels=labels,
            restart_policy={"Name": "no"},
        )

    async def _cleanup_when_done(self, job_id: str) -> None:
        # Poll job status in Redis; when terminal, stop/remove containers.
        terminal = {"completed", "cancelled", "failed"}
        while True:
            job = await self.coordinator.get_job(job_id)
            status = (job or {}).get("status") or ""
            if status in terminal:
                break
            await asyncio.sleep(3)

        if self.settings.cleanup_delay_seconds > 0:
            await asyncio.sleep(self.settings.cleanup_delay_seconds)

        prefix = f"agentfleet-{job_id[:8]}-"
        for c in self.docker_client.containers.list(all=True):
            if not c.name.startswith(prefix):
                continue
            try:
                c.remove(force=True)
            except Exception:
                continue

        print(
            f"[orchestrator] cleaned crew job_id={job_id} status={status}",
            flush=True,
        )


async def main() -> None:
    settings = OrchestratorSettings.from_env()
    orch = Orchestrator(settings)
    await orch.start()

File: packages/agents/src/orchestrator/orchestrator.ts
import { setTimeout } from "node:timers/promises";
import os from "node:os";
import Docker from "dockerode";
import {
  RedisCoordinator,
  AgentMessage,
  WorkerRequestResponse,
  utcNow,
} from "@agent-fleet/shared";

// ---------------------------------------------------------------------------
// Settings
// ---------------------------------------------------------------------------

interface OrchestratorSettings {
  redisUrl: string;
  agentsImage: string;
  dockerNetwork: string;
  repoVolume: string;
  maxWorkersPerJob: number;
  cleanupDelaySeconds: number;
}

function settingsFromEnv(): OrchestratorSettings {
  return {
    redisUrl: process.env.REDIS_URL ?? "redis://redis:6379/0",
    agentsImage: process.env.AGENTS_IMAGE ?? "agentfleet-agents:latest",
    dockerNetwork: process.env.DOCKER_NETWORK ?? "agentfleet_default",
    repoVolume: process.env.REPO_VOLUME ?? "agentfleet_repo",
    maxWorkersPerJob: parseInt(
      process.env.MAX_WORKERS_PER_JOB ?? "6",
      10,
    ),
    cleanupDelaySeconds: parseFloat(
      process.env.CLEANUP_DELAY_SECONDS ?? "0",
    ),
  };
}

// ---------------------------------------------------------------------------
// Orchestrator
// ---------------------------------------------------------------------------

class Orchestrator {
  private readonly settings: OrchestratorSettings;
  private readonly coordinator: RedisCoordinator;
  private readonly docker: Docker;
  private readonly hostname: string;

  constructor(settings: OrchestratorSettings) {
    this.settings = settings;
    this.coordinator = new RedisCoordinator(settings.redisUrl);
    this.docker = new Docker({ socketPath: "/var/run/docker.sock" });
    this.hostname = os.hostname();
  }

  async start(): Promise<void> {
    await this.coordinator.ensureConsumerGroup("orchestrators");
    this.workerRequestLoop();

    while (true) {
      const messages = await this.coordinator.readJobs(
        "orchestrators",
        `orch-${this.hostname.slice(0, 8)}`,
        5000,
      );
      if (!messages) continue;

      for (const [, entries] of messages) {
        for (const [entryId, fields] of entries) {
          const raw =
            typeof fields === "object" && !Array.isArray(fields)
              ? (fields as Record<string, string>).data
              : Array.isArray(fields)
                ? fields[fields.indexOf("data") + 1]
                : undefined;
          if (!raw) {
            await this.coordinator.ackJob("orchestrators", entryId);
            continue;
          }

          let job: Record<string, unknown>;
          try {
            job = JSON.parse(raw) as Record<string, unknown>;
          } catch {
            await this.coordinator.ackJob("orchestrators", entryId);
            continue;
          }

          try {
            await this.spawnLeader(job);
            await this.coordinator.ackJob("orchestrators", entryId);
          } catch (err: unknown) {
            const jobId = String(job.job_id ?? job.id ?? "");
            if (jobId) {
              await this.coordinator.publishJobEvent(jobId, {
                type: "orchestrator_error",
                error: err instanceof Error ? err.message : String(err),
              });
            }
            await this.coordinator.ackJob("orchestrators", entryId);
          }
        }
      }
    }
  }

  // -------------------------------------------------------------------------
  // Spawn leader
  // -------------------------------------------------------------------------

  private async spawnLeader(job: Record<string, unknown>): Promise<void> {
    const jobId = String(job.job_id ?? job.id ?? "");
    if (!jobId) throw new Error("job_id missing");

    const prompt = String(job.prompt ?? "");
    await this.coordinator.upsertJob(jobId, {
      job_id: jobId,
      prompt,
      status: "queued",
      orchestrated: true,
    });

    const leaderId = `leader-${jobId.slice(0, 8)}`;
    console.log(
      `[orchestrator] spawning leader for job_id=${jobId}`,
    );

    const baseEnv = this.buildBaseEnv(jobId);
    await this.runContainer({
      name: `agentfleet-${jobId.slice(0, 8)}-leader`,
      env: { ...baseEnv, AGENT_ROLE: "leader", AGENT_ID: leaderId },
      labels: { "agentteam.job_id": jobId, "agentteam.role": "leader" },
    });

    await this.coordinator.publishJobEvent(jobId, {
      type: "leader_spawned",
      job_id: jobId,
      leader_id: leaderId,
    });

    this.cleanupWhenDone(jobId);
  }

  // -------------------------------------------------------------------------
  // Worker request loop
  // -------------------------------------------------------------------------

  private async workerRequestLoop(): Promise<void> {
    while (true) {
      try {
        const request = await this.coordinator.waitForWorkerRequest(5);
        if (!request) continue;

        const jobId = request.job_id;
        const leaderId = request.leader_id;
        const requested = request.requested_count;
        const granted = Math.min(
          requested,
          this.settings.maxWorkersPerJob,
        );

        console.log(
          `[orchestrator] spawning ${granted} workers for job_id=${jobId} (requested=${requested})`,
        );

        const baseEnv = this.buildBaseEnv(jobId);
        const workerIds: string[] = [];

        for (let i = 0; i < granted; i++) {
          const workerId = `worker-${jobId.slice(0, 8)}-${i}`;
          workerIds.push(workerId);
          await this.runContainer({
            name: `agentfleet-${jobId.slice(0, 8)}-worker-${i}`,
            env: {
              ...baseEnv,
              AGENT_ROLE: "worker",
              AGENT_ID: workerId,
            },
            labels: {
              "agentteam.job_id": jobId,
              "agentteam.role": "worker",
            },
          });
        }

        // Notify leader
        const response = WorkerRequestResponse.parse({
          job_id: jobId,
          granted_count: granted,
          worker_ids: workerIds,
          timestamp: utcNow(),
        });
        const msg = AgentMessage.parse({
          from_agent: "orchestrator",
          text: JSON.stringify(response),
          summary: `Granted ${granted} workers`,
        });
        await this.coordinator.sendMessage(jobId, leaderId, msg);

        await this.coordinator.publishJobEvent(jobId, {
          type: "workers_spawned",
          job_id: jobId,
          leader_id: leaderId,
          granted,
          worker_ids: workerIds,
        });
      } catch (err: unknown) {
        console.log(
          `[orchestrator] worker_request_loop error: ${err instanceof Error ? err.message : err}`,
        );
        await setTimeout(1000);
      }
    }
  }

  // -------------------------------------------------------------------------
  // Container management
  // -------------------------------------------------------------------------

  private buildBaseEnv(jobId: string): Record<string, string> {
    return {
      REDIS_URL: this.settings.redisUrl,
      AGENT_SDK_MODE: process.env.AGENT_SDK_MODE ?? "live",
      MODEL: process.env.MODEL ?? "",
      ANTHROPIC_API_KEY: process.env.ANTHROPIC_API_KEY ?? "",
      OPENAI_API_KEY: process.env.OPENAI_API_KEY ?? "",
      GOOGLE_API_KEY: process.env.GOOGLE_API_KEY ?? "",
      DRY_RUN_SLEEP_SECONDS:
        process.env.DRY_RUN_SLEEP_SECONDS ?? "0",
      JOB_ID: jobId,
      JOB_MODE: "ephemeral",
      MAX_WORKERS_PER_JOB: String(this.settings.maxWorkersPerJob),
    };
  }

  private async runContainer(opts: {
    name: string;
    env: Record<string, string>;
    labels: Record<string, string>;
  }): Promise<void> {
    // Remove existing container if present
    try {
      const existing = this.docker.getContainer(opts.name);
      await existing.remove({ force: true });
    } catch {
      // does not exist
    }

    const envList = Object.entries(opts.env).map(
      ([k, v]) => `${k}=${v}`,
    );

    const container = await this.docker.createContainer({
      Image: this.settings.agentsImage,
      name: opts.name,
      Env: envList,
      Labels: opts.labels,
      HostConfig: {
        NetworkMode: this.settings.dockerNetwork,
        Binds: [`${this.settings.repoVolume}:/repo:rw`],
        RestartPolicy: { Name: "" },
      },
    });
    await container.start();
  }

  private async cleanupWhenDone(jobId: string): Promise<void> {
    const terminal = new Set(["completed", "cancelled", "failed"]);
    let status = "";
    while (true) {
      const job = await this.coordinator.getJob(jobId);
      status = String(job?.status ?? "");
      if (terminal.has(status)) break;
      await setTimeout(3000);
    }

    if (this.settings.cleanupDelaySeconds > 0) {
      await setTimeout(this.settings.cleanupDelaySeconds * 1000);
    }

    const prefix = `agentfleet-${jobId.slice(0, 8)}-`;
    const containers = await this.docker.listContainers({ all: true });
    for (const info of containers) {
      const names = info.Names ?? [];
      if (!names.some((n) => n.replace("/", "").startsWith(prefix)))
        continue;
      try {
        await this.docker.getContainer(info.Id).remove({ force: true });
      } catch {
        continue;
      }
    }
    console.log(
      `[orchestrator] cleaned crew job_id=${jobId} status=${status}`,
    );
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

export async function main(): Promise<void> {
  const settings = settingsFromEnv();
  const orch = new Orchestrator(settings);
  await orch.start();
}
